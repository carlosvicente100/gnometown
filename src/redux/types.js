// gnomes
export const fetchRequested = 'FETCH_REQUESTED';
export const fetchRequestedGnomes = 'FETCH_REQUESTED_GNOMES';
export const fetchPendingGnomes = 'FETCH_PENDING_GNOMES';
export const fetchSucceededGnomes = 'FETCH_SUCCEEDED_GNOMES';
export const fetchFailedGnomes = 'FETCH_FAILED_GNOMES';

//filters
export const setAgeFromTo = 'SET_BREWED_FROM_TO';
export const setAgeFrom = 'SET_BREWED_FROM';
export const setAgeTo = 'SET_BREWED_TO';
