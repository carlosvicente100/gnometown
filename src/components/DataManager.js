import React from 'react'
import { useSelector } from 'react-redux'
import styled from 'styled-components'

const Loader = styled.div`
  width: 100%;
  height: 100%;
  display: flex;
  justify-content: center;
  align-items: center;
`

const Error = styled.div`
  margin-top: 50px;
  background-color: red;
  width: 100%;
  height: 50px;
  display: flex;
  justify-content: center;
  align-items: center;

  label {
    font-family: Helvetica;
    color: #ffffff;
    font-size: 15px;
  }
`
const DataManager = ({ children }) => {
  const { error, pending } = useSelector((state) => state.gnomesReducer)
  return (
    <>
      {pending && (
        <Loader>
          <img alt="loading" src="https://createwebsite.net/wp-content/uploads/2015/09/Display-Loading.gif" />
        </Loader>
      )}
      {!pending && error && (
        <Error>
          <label>We're having problems Searching Gnomes, try it in a few minutes</label>
        </Error>
      )}
      {!pending && !error && <>{children}</>}
    </>
  )
}

export default DataManager
