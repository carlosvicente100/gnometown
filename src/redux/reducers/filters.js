import { fetchSetJobList, fetchSetName, fetchResetFilters } from '../types'

export const initialState = {
  jobList: [],
  name: ''
}

export default function reducer(state = initialState, action) {
  switch (action.type) {
    case fetchSetJobList:
      return {
        ...state,
        jobList: action.data.jobs
      }
    case fetchSetName:
      return {
        ...state,
        name: action.data.name
      }
    case fetchResetFilters:
      return {
        ...initialState
      }
    default:
      return {
        ...state
      }
  }
}
