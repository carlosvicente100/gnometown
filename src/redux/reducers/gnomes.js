import { fetchPendingGnomes, fetchSucceededGnomes, fetchFailedGnomes } from '../types'

export const initialState = {
  pending: false,
  error: null,
  townName: '',
  gnomes: []
}

export default function reducer(state = initialState, action) {
  switch (action.type) {
    case fetchPendingGnomes:
      return {
        ...state,
        pending: true
      }
    case fetchSucceededGnomes:
      const townName = Object.keys(action.data)[0]
      return {
        ...state,
        pending: false,
        townName,
        gnomes: action.data[townName]
      }
    case fetchFailedGnomes:
      return {
        ...state,
        pending: false,
        error: true
      }
    default:
      return {
        ...state
      }
  }
}
