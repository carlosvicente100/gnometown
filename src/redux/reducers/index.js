import { combineReducers } from 'redux';
 import gnomesReducer from './gnomes';
import filtersReducer from './filters';

export default combineReducers({
  gnomesReducer,
  filtersReducer,
});
