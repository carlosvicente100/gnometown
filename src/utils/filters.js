export const getJobsByUsers = (users) => {
  //   console.log('getJobsByUsers', users)
  if (users.length > 0) {
    const profesionList = users
      .map((user) => user.professions)
      .reduce((firstArray, secondArray) => [...firstArray, ...secondArray], [])
    return profesionList.filter((job, index) => profesionList.indexOf(job) === index)
  } else {
    return []
  }
}

export const DEFAULT_VALUE = ''
export const JOBS = 'Jobs'

export const makePagination = (itemsFiltered) => {
  return itemsFiltered.reduce((result, value, index) => {
    if (index % 40 !== 0) {
      result[result.length - 1].push(value)
    } else {
      result.push([value])
    }
    return result
  }, [])
}

export const filterByJob = (jobs, users) => {
  let gnomesWithJobs = users.filter((user) => {
    if (jobs.every((job) => user.professions.includes(job))) {
      return user
    } else {
      return false
    }
  })
  return gnomesWithJobs
}
