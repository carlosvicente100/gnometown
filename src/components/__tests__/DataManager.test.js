import React from 'react'
import { Provider } from 'react-redux'
import { createStore } from 'redux'
import { render as rtlRender } from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'
import 'jest-styled-components'

import DataManager from '../DataManager'

import { initialState as initialStateGnomes } from '../../redux/reducers/gnomes'
import { initialState as initialStateFilters } from '../../redux/reducers/filters'
import reducer from '../../redux/reducers'

const initialReducerState = {
  gnomesReducer: initialStateGnomes,
  filtersReducer: initialStateFilters
}
function render(ui, { initialState, store = createStore(reducer, initialState), ...renderOptions } = {}) {
  function Wrapper({ children }) {
    return <Provider store={store}>{children}</Provider>
  }
  return rtlRender(ui, { wrapper: Wrapper, ...renderOptions })
}

describe('DataManager Component', () => {
  test('DataManager Snapshot', () => {
    expect(
      render(
        <DataManager>
          <h1>correct data</h1>
        </DataManager>,
        {
          initialState: { gnomesReducer: { pending: false, error: false } }
        }
      ).baseElement
    ).toMatchSnapshot()

    expect(
      render(
        <DataManager>
          <h1>data correct</h1>
        </DataManager>,
        {
          initialState: { gnomesReducer: { pending: true, error: false } }
        }
      ).baseElement
    ).toMatchSnapshot()

    expect(
      render(
        <DataManager>
          <h1>data correct</h1>
        </DataManager>,
        {
          initialState: { gnomesReducer: { pending: false, error: true } }
        }
      ).baseElement
    ).toMatchSnapshot()
  })

  test('DataManager should show loader if there is pending', () => {
    const { container } = render(
      <DataManager>
        <h1>data correct</h1>
      </DataManager>,
      {
        initialState: { gnomesReducer: { pending: true, error: false } }
      }
    )
    const loadingImg = container.querySelector('img')
    expect(loadingImg.getAttribute('alt')).toBe('loading')
    expect(loadingImg.getAttribute('src')).toBe(
      'https://createwebsite.net/wp-content/uploads/2015/09/Display-Loading.gif'
    )
  })

  test('DataManager should show children if is not pending and there is no error', () => {
    const { container } = render(
      <DataManager>
        <h1>data correct</h1>
      </DataManager>,
      {
        initialState: { gnomesReducer: { pending: false, error: false } }
      }
    )
    expect(container.querySelector('h1').innerHTML).toBe('data correct')
  })

  test('DataManager should show error message if there is error', () => {
    const { container } = render(
      <DataManager>
        <h1>data correct</h1>
      </DataManager>,
      {
        initialState: { gnomesReducer: { pending: false, error: true } }
      }
    )
    expect(container.querySelector('label').innerHTML).toBe(
      "We're having problems Searching Gnomes, try it in a few minutes"
    )
  })
})
