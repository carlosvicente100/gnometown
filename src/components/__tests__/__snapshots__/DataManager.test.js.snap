// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataManager Component DataManager Snapshot 1`] = `
<body>
  <div>
    <h1>
      correct data
    </h1>
  </div>
</body>
`;

exports[`DataManager Component DataManager Snapshot 2`] = `
<body>
  <div>
    <h1>
      correct data
    </h1>
  </div>
  .c0 {
  width: 100%;
  height: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

<div>
    <div
      class="c0"
    >
      <img
        alt="loading"
        src="https://createwebsite.net/wp-content/uploads/2015/09/Display-Loading.gif"
      />
    </div>
  </div>
</body>
`;

exports[`DataManager Component DataManager Snapshot 3`] = `
<body>
  <div>
    <h1>
      correct data
    </h1>
  </div>
  <div>
    <div
      class="sc-AxjAm kSkoow"
    >
      <img
        alt="loading"
        src="https://createwebsite.net/wp-content/uploads/2015/09/Display-Loading.gif"
      />
    </div>
  </div>
  .c0 {
  margin-top: 50px;
  background-color: red;
  width: 100%;
  height: 50px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c0 label {
  font-family: Helvetica;
  color: #ffffff;
  font-size: 15px;
}

<div>
    <div
      class="c0"
    >
      <label>
        We're having problems Searching Gnomes, try it in a few minutes
      </label>
    </div>
  </div>
</body>
`;
